src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs

```diff
@@ -94,28 +114,29 @@ protected internal XmlDocument(XmlImplementation imp) : base()
             _implementation = imp;
             _domNameTable = new DomNameTable(this);
 
-            // force the following string instances to be default in the nametable
-            XmlNameTable nt = this.NameTable;
-            nt.Add(string.Empty);
-            strDocumentName = nt.Add("#document");
-            strDocumentFragmentName = nt.Add("#document-fragment");
-            strCommentName = nt.Add("#comment");
-            strTextName = nt.Add("#text");
-            strCDataSectionName = nt.Add("#cdata-section");
-            strEntityName = nt.Add("#entity");
-            strID = nt.Add("id");
-            strNonSignificantWhitespaceName = nt.Add("#whitespace");
-            strSignificantWhitespaceName = nt.Add("#significant-whitespace");
-            strXmlns = nt.Add("xmlns");
-            strXml = nt.Add("xml");
-            strSpace = nt.Add("space");
-            strLang = nt.Add("lang");
-            strReservedXmlns = nt.Add(XmlReservedNs.NsXmlNs);
-            strReservedXml = nt.Add(XmlReservedNs.NsXml);
-            strEmpty = nt.Add(String.Empty);
+            strXmlns = Xmlns;
+            strXml = Xml;
+            strReservedXmlns = XmlReservedNs.NsXmlNs;
+            strReservedXml = XmlReservedNs.NsXml;
             baseURI = String.Empty;
-
             objLock = new object();
+
+            NameTable nt = imp.NameTable as NameTable;
+            if (nt != null)
+            {
+                for (uint i = 0; i < s_nameTableSeeds.Length; ++i)
+                {
+                   nt.AddEntry(s_nameTableSeeds[i].key, s_nameTableSeeds[i].hash);
```

Comment:
AddEntry appears to assume that the entry doesn't exist.  What happens if someone passes the already-filled NameTable from one XmlDocument to another:
```C#
var d1 = new XmlDocument();
var d2 = new XmlDocument(d1.NameTable);
```
? Won't this end up duplicating the entry, whereas previously Add would have weeded out the duplicates?

Reply:
You're right, this will break.

Reply:
This actually means that I can't remove all those internal fields: the issue is if someone already added those to the table before passing it to XmlDocument and they are using reference equals to check if it is the same string, then I can't safely rely on the const. It is a corner case but broken in current form.